docs/source/tutorial.rst

```diff
@@ -0,0 +1,659 @@
+This setup function is automatically called before each test.
+Therefore, our test could execute `project.sh` directly, without using a (relative) path.
+
+.. important::
+
+    The `setup` function will be called before each individual test in the file. 
+    Each file can only define one setup function for all tests in the file.
+    However, the setup functions can differ between different files.
+
+Dealing with output
+-------------------
+
+Okay, we have a green test but our executable does not anything useful.
+To keep things simple, let us start with an error message. Our new `src/project.sh` now reads:
+
+.. code-block:: bash
+
+    #!/usr/bin/env bash
+
+    echo "Welcome to our project!"
+
+    echo "NOT IMPLEMENTED!" >&2
+    exit 1
+
+And gives is this test output:
+
+.. code-block:: console
+
+    $ ./test/bats/bin/bats test/test.bats 
+     ✗ can run our script
+       (in test file test/test.bats, line 11)
+         `project.sh' failed
+       Welcome to our project!
+       NOT IMPLEMENTED!
+
+    1 test, 1 failure
+
+Okay, our test failed again, because we now exit with 1 instead of 0.
+Additionally, we see the stdout and stderr of the failing program.
+
+Our goal now is to retarget our test and check that we get the welcome message.
+bats-assert gives us some help with this, so we should now load it (and its dependency bats-support),
+so we change `test/test.bats` to
+
+.. code-block:: bash
+
+    setup() {
+        load 'test_helper/bats-support/load'
+        load 'test_helper/bats-assert/load'
+        # ... the remaining setup is unchanged
+
+        # get the containing directory of this file
+        # use $BATS_TEST_FILENAME instead of ${BASH_SOURCE[0]} or $0,
+        # as those will point to the bats executable's location or the preprocessed file respectively
+        DIR="$( cd "$( dirname "$BATS_TEST_FILENAME" )" >/dev/null 2>&1 && pwd )"
+        # make executables in src/ visible to PATH
+        PATH="$DIR/../src:$PATH"
+    }
+
+    @test "can run our script" {
+        run project.sh # notice `run`!
+        assert_output 'Welcome to our project!'
+    }
+
+which gives us the following test output:
+
+.. code-block:: console
+
+    $ LANG=C ./test/bats/bin/bats test/test.bats 
+     ✗ can run our script
+       (from function `assert_output' in file test/test_helper/bats-assert/src/assert_output.bash, line 194,
+        in test file test/test.bats, line 14)
+         `assert_output 'Welcome to our project!'' failed
+    
+       -- output differs --
+       expected (1 lines):
+         Welcome to our project!
+       actual (2 lines):
+         Welcome to our project!
+         NOT IMPLEMENTED!
+       --
+    
+
+    1 test, 1 failure
+
+The first change in this output is the failure description. We now fail on assert_output instead of the call itself.
+We prefixed our call to `project.sh` with `run`, which is an internal bats function that executes the command it gets passed as parameters.
```

Comment:
```suggestion
We prefixed our call to `project.sh` with `run`, which is a bats function that executes the command it gets passed as parameters.
```

Since we expect users to use it, I don't think it should be described as an internal function.